src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,171 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Index " + index + "out of size " + size);
+        } else {
+            Node<T> next = takeElementByIndex(index);
+            if (next == null || index == size) {
+                add(value);
+                return;
+            }
+            Node<T> prev = next.prev;
+            Node<T> newNode = new Node<>(prev, value, next);
+            next.prev = newNode;
+            if (prev == null) {
+                head = newNode;
+            } else {
+                prev.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T values : list) {
+            add(values);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> elementByIndex = takeElementByIndex(index);
```

Comment:
Move the index checking to the takeElementByIndex method

Reply:
I use takeElementByIndex method in the public void add(T value, int index) method and there are other criteria for checking index

Reply:
In `get(int index)` and `takeElementByIndex ` logic for checking index is the same. So, add checking inside `takeElementByIndex` and remove it here.